File: archive/j/java/MaximumArrayRotation.java
import java.util.Arrays;

public class MaximumArrayRotation {
    public static void main(String[] args) {
        int[] input = parse(args);
        System.out.println(maximumRotationSum(input));
    }

    private static void usage() {
        System.err.println("Usage: please provide a list of integers (e.g. \"8, 3, 1, 2\")");
        System.exit(1);
    }

    private static int[] parse(String[] args) {
        if (args.length != 1 || args[0] == null || args[0].isBlank()) {
            usage();
        }

        try {
            int[] values = Arrays.stream(args[0].split(","))
                    .map(String::trim)
                    .mapToInt(Integer::parseInt)
                    .toArray();

            if (values.length == 0) {
                usage();
            }

            return values;

        } catch (NumberFormatException e) {
            usage();
            return new int[0]; // unreachable
        }
    }

    private static long maximumRotationSum(int[] values) {
        int n = values.length;
        if (n == 0) {
            usage();
        }

        long sum = 0;
        long currentSum = 0;

        for (int i = 0; i < n; i++) {
            sum += values[i];
            currentSum += (long) values[i] * i;
        }

        long maxSum = currentSum;

        for (int i = 1; i < n; i++) {
            currentSum += sum - (long) n * values[n - i];
            maxSum = Math.max(maxSum, currentSum);
        }

        return maxSum;
    }
}

File: archive/j/javascript/maximum-array-rotation.js
"use strict";

const USAGE = 'Usage: please provide a list of integers (e.g. "8, 3, 1, 2")';

const parse = (s) =>
  s
    .split(",")
    .filter((x) => x.trim() !== "")
    .map((x) => Number(x.trim()));

const maxArrayRotation = (arr) => {
  const n = arr.length;
  if (n === 0) return 0;

  let currentSum = 0;
  let totalSum = 0;

  for (let i = 0; i < n; i++) {
    totalSum += arr[i];
    currentSum += i * arr[i];
  }

  let maxSum = currentSum;

  for (let k = 1; k < n; k++) {
    const droppedValue = arr[n - k];
    currentSum = currentSum + totalSum - n * droppedValue;

    if (currentSum > maxSum) {
      maxSum = currentSum;
    }
  }

  return maxSum;
};

const run = () => {
  const [, , input] = process.argv;

  if (!input) {
    console.error(USAGE);
    process.exit(1);
  }

  const arr = parse(input);

  if (arr.length === 0 || arr.some(Number.isNaN)) {
    console.error(USAGE);
    process.exit(1);
  }

  console.log(maxArrayRotation(arr));
};

run();
